### DISCUSSION TITLE
Deploy to IIS (or how to ask NextJS to accept a string for port)

### DISCUSSION THREAD

**Discussion by mcmillhj-wta** [vercel/next.js#22922] [Help]
TL;DR is there a way to ask Next to accept a string for PORT without making a custom server and losing out on Automatic Static Optimization? Alternatively, if we create a custom server but still forward all traffic to `handle` do we lose out on Automatic Static Optimization? 

Currently, our team is in the process of deploying our NextJS application to IIS on a Windows App Service within Azure. There are a few problems that we have encountered along the way:

1. iisnode doesn't allow a start command such as `next start`, instead you have to point to a file. We created a small script called `start.js` that wraps `node_modules/next/dist/cli/next-start.js` and passes in the allowed hostname and port parameters:
      ```
       <!-- web.config snippet -->
        <handlers>
          <add name="iisnode" path="start.js" verb="*" modules="iisnode" />
        </handlers>
      ```

2. iisnode uses a named pipe as the PORT for forwarding traffic from ports 80/443 to the underlying NodeJS application. We are currently stuck on this because NextJS validates that the supplied port is a number and if it isn't NextJS uses the default value of `3000` 

3. If we create a custom server to pass through the string port from (2), but route all of the traffic to NextJS do we still lose out on automatic static optimization? 

    ```
    /* eslint-disable @typescript-eslint/no-var-requires */
    const { createServer } = require('http');
    const next = require('next');
    
    const port = process.env.PORT || 3000;
    const dev = process.env.NODE_ENV !== 'production';
    const app = next({ dev });
    const handle = app.getRequestHandler();
    
    app.prepare().then(() => {
      return createServer((req, res) => {
        handle(req, res);
      }).listen(port, (err) => {
        if (err) throw err;
        // eslint-disable-next-line no-console
        console.log(`Listening on http://localhost:${port}`);
      });
    });
    ```

**Comment by leerob** [comment]
When running `next start`, you can add `next start -p 4000` to change the port number.

**Reply by mcmillhj-wta** [reply]
In point 1 above I describe why we can't use `next start` in this case. In point 2, I describe why a port _number_ also doesn't work in this case.

**Reply by mcmillhj-wta** [reply]
@leerob any other thoughts on our setup and if we'll have any complications?

**Comment by AnweshGangula** [comment]
@mcmillhj-wta,  were you able to find any solution to this?
I'm not entirely sure if this works, but can you checkout the answer in the [following link](https://stackoverflow.com/a/63098601/6908282).
